Generate HMR initializer code

crisbeto opened the pull request:
Includes a couple of commits that enable the compiler to output calls that will enable HMR for component classes, as well as fix for file paths in `setClassDebugInfo`.

### refactor(compiler): add logic to generate the HMR initializer
Adds the logic that will generate the `import` expression that will initializer HMR for a specific component.

### refactor(compiler-cli): output HMR initializer code
Adds the logic to the compiler that will output the HMR initializer code for each component, if enabled.

### fix(compiler-cli): incorrectly generating relative file paths on case-insensitive platforms

We're using `path.relative` to compute a relative path between a `SourceFile` and the one of the `rootDirs`. The problem is that the `rootDirs` get passed through `getCanonicalFileName` which lowercases the path in some platforms, while `SourceFile.fileName` is always case-insensitive. This yields a path outside of the project which we were ignoring.

This change passes the `SourceFile.fileName` before passing it through `path.relative` to ensure that we get a valid result.

josephperrott reviewed (commented):
LGTM

Reviewed-for: bazel

josephperrott reviewed (approved):
LGTM

(forgot to choose approve)

Reviewed-for: bazel

crisbeto committed:
refactor(compiler): add logic to generate the HMR initializer

Adds the logic that will generate the `import` expression that will initializer HMR for a specific component.

crisbeto committed:
refactor(compiler-cli): output HMR initializer code

Adds the logic to the compiler that will output the HMR initializer code for each component, if enabled.

crisbeto committed:
fix(compiler-cli): incorrectly generating relative file paths on case-insensitive platforms

We're using `path.relative` to compute a relative path between a `SourceFile` and the one of the `rootDirs`. The problem is that the `rootDirs` get passed through `getCanonicalFileName` which lowercases the path in some platforms, while `SourceFile.fileName` is always case-insensitive. This yields a path outside of the project which we were ignoring.

This change passes the `SourceFile.fileName` before passing it through `path.relative` to ensure that we get a valid result.

devversion commented:
This PR was merged into the repository by commit fb44323c51da5a86853aafd8a70ce0c25d6c0d7f.

The changes were merged into the following branches: main
